# Hamoye Stage B Quiz
## Appliances Energy Prediction Dataset
### Dataset Description
The dataset for the remainder of this quiz is the Appliances Energy Prediction data. The data set is at 10 min for about 4.5 months. The house temperature and humidity conditions were monitored with a ZigBee wireless sensor network. Each wireless node transmitted the temperature and humidity conditions around 3.3 min. Then, the wireless data was averaged for 10 minutes periods. The energy data was logged every 10 minutes with m-bus energy meters. Weather from the nearest airport weather station (Chievres Airport, Belgium) was downloaded from a public data set from Reliable Prognosis (rp5.ru), and merged together with the experimental data sets using the date and time column. Two random variables have been included in the data set for testing the regression models and to filter out non predictive attributes (parameters). The attribute information can be seen below.

Attribute Information:
Date, time year-month-day hour:minute:second
Appliances, energy use in Wh
lights, energy use of light fixtures in the house in Wh
T1, Temperature in kitchen area, in Celsius
RH_1, Humidity in kitchen area, in %
T2, Temperature in living room area, in Celsius
RH_2, Humidity in living room area, in %
T3, Temperature in laundry room area
RH_3, Humidity in laundry room area, in %
T4, Temperature in office room, in Celsius
RH_4, Humidity in office room, in %
T5, Temperature in bathroom, in Celsius
RH_5, Humidity in bathroom, in %
T6, Temperature outside the building (north side), in Celsius
RH_6, Humidity outside the building (north side), in %
T7, Temperature in ironing room , in Celsius
RH_7, Humidity in ironing room, in %
T8, Temperature in teenager room 2, in Celsius
RH_8, Humidity in teenager room 2, in %
T9, Temperature in parents room, in Celsius
RH_9, Humidity in parents room, in %
To, Temperature outside (from Chievres weather station), in Celsius
Pressure (from Chievres weather station), in mm Hg
RH_out, Humidity outside (from Chievres weather station), in %
Wind speed (from Chievres weather station), in m/s
Visibility (from Chievres weather station), in km
Tdewpoint (from Chievres weather station), Â °C
rv1, Random variable 1, nondimensiona
rv2, Random variable 2, nondimensional

### Instructions
To answer some questions, you will need to normalize the dataset using the MinMaxScaler after removing the following columns: [“date”, “lights”]. The target variable is “Appliances”. Use a 70-30 train-test set split with a  random state of 42 (for reproducibility). Run a multiple linear regression using the training set and evaluate your model on the test set.

In [38]:
#Import numpy for numerical computation
import numpy as np
#import pandas for data processing and cleaning
import pandas as pd
energy_data = pd.read_csv (r'C:\Users\Nichel\Anaconda2\Lib\site-packages\pandas\energydata_complete.csv')
#rename columns
column_names = {'T1': 'kitchen_temp', 'RH_1': 'kitchen_hum', 'T2': 'living_room_temp',
              'RH_2': 'living_room_hum', 'T3': 'laundry_room_temp' , 'RH_3': 'laundry_room_hum', 'T4': 'office_room_temp' , 'RH_4': 'office_room_hum', 
               'T5': 'bath_room_temp' , 'RH_5': 'bath_room_hum', 'T6': 'outside_building_temp' , 'RH_6': 'outside_building_hum', 'T7': 'ironing_room_temp' , 'RH_3': 'ironing_room_hum',
              'T8': 'teenager_room2_temp' , 'RH_8': 'teenager_room2_hum', 'T9': 'parent_room_temp' , 'RH_9': 'parent_room_hum', 'T_out': 'Outside_station_temp' , 'RH_out': 'outside_station_hum'}
energy_data = energy_data.rename(columns = column_names)

In [2]:
energy_data.shape 

(19735, 29)

### Simple Linear Regression Model

In [3]:
df = pd.DataFrame(energy_data, columns = [ 'living_room_temp', 'outside_building_temp' ])
x = df[['living_room_temp']]
y = df['outside_building_temp']

In [4]:
from sklearn.linear_model import LinearRegression
linear_model = LinearRegression()
linear_model.fit(x, y)

LinearRegression(copy_X=True, fit_intercept=True, n_jobs=None,
         normalize=False)

In [5]:
predicted_value =linear_model.predict(x)

In [6]:
from sklearn.metrics import r2_score
round(r2_score(y, predicted_value), 2)

0.64

#### Question 12
From the dataset, fit a linear model on the relationship between the temperature in the living room in Celsius(x =T2) and the temperature outside the building(y = T6). What is the R^2 value in two D.P
#### Answer = 0.64

### Multiple Linear Regression Model

In [7]:
energy_data_revised = energy_data.drop(['date','lights'], axis =1)
energy_data_revised

Appliances  kitchen_temp  kitchen_hum  living_room_temp  \
0              60     19.890000    47.596667         19.200000   
1              60     19.890000    46.693333         19.200000   
2              50     19.890000    46.300000         19.200000   
3              50     19.890000    46.066667         19.200000   
4              60     19.890000    46.333333         19.200000   
5              50     19.890000    46.026667         19.200000   
6              60     19.890000    45.766667         19.200000   
7              60     19.856667    45.560000         19.200000   
8              60     19.790000    45.597500         19.200000   
9              70     19.856667    46.090000         19.230000   
10            230     19.926667    45.863333         19.356667   
11            580     20.066667    46.396667         19.426667   
12            430     20.133333    48.000000         19.566667   
13            250     20.260000    52.726667         19.730000   
14            100     20.426667    55.893333         19.856667   
15            100     20.566667    53.893333         20.033333   
16             90     20.730000    52.660000         20.166667   
17             70     20.856667    53.660000         20.200000   
18             80     20.890000    51.193333         20.200000   
19            140     20.890000    49.800000         20.200000   
20            120     20.890000    48.433333         20.200000   
21            190     20.963333    47.633333         20.260000   
22            110     21.033333    47.063333         20.290000   
23            110     21.100000    46.596667         20.356667   
24            110     21.133333    46.060000         20.426667   
25            110     21.200000    45.800000         20.500000   
26            100     21.290000    45.900000         20.533333   
27            100     21.356667    45.826667         20.666667   
28            100     21.390000    45.690000         20.700000   
29            100     21.500000    45.333333         20.700000   
...           ...           ...          ...               ...   
19705         280     25.033333    48.363333         26.528571   
19706         240     25.166667    48.156667         26.600000   
19707         250     25.323333    47.930000         26.600000   
19708         220     25.390000    47.656667         26.600000   
19709         230     25.500000    47.133333         26.600000   
19710         190     25.500000    47.060000         26.540000   
19711         160     25.600000    46.990000         26.512500   
19712          80     25.600000    46.730000         26.437143   
19713          60     25.566667    46.633333         26.370000   
19714          60     25.500000    46.360000         26.318571   
19715          60     25.500000    46.060000         26.350000   
19716          60     25.500000    45.933333         26.277143   
19717          70     25.500000    45.760000         26.200000   
19718          80     25.500000    45.626667         26.171429   
19719          80     25.500000    45.590000         26.100000   
19720          70     25.500000    45.522500         26.100000   
19721         100     25.500000    45.633333         26.080000   
19722         100     25.500000    45.730000         26.000000   
19723          90     25.500000    45.790000         26.000000   
19724         100     25.500000    45.933333         26.000000   
19725         220     25.426667    46.060000         26.000000   
19726         180     25.500000    46.530000         26.000000   
19727         120     25.500000    47.456667         26.000000   
19728         110     25.600000    47.193333         25.968571   
19729          90     25.533333    46.860000         25.978000   
19730         100     25.566667    46.560000         25.890000   
19731          90     25.500000    46.500000         25.754000   
19732         270     25.500000    46.596667         25.628571   
19733         420     25.500000    46.990000   

In [8]:
#show the summary statistics
energy_data_revised.describe()

Appliances  kitchen_temp   kitchen_hum  living_room_temp  \
count  19735.000000  19735.000000  19735.000000      19735.000000   
mean      97.694958     21.686571     40.259739         20.341219   
std      102.524891      1.606066      3.979299          2.192974   
min       10.000000     16.790000     27.023333         16.100000   
25%       50.000000     20.760000     37.333333         18.790000   
50%       60.000000     21.600000     39.656667         20.000000   
75%      100.000000     22.600000     43.066667         21.500000   
max     1080.000000     26.260000     63.360000         29.856667   

       living_room_hum  laundry_room_temp  ironing_room_hum  office_room_temp  \
count     19735.000000       19735.000000      19735.000000      19735.000000   
mean         40.420420          22.267611         39.242500         20.855335   
std           4.069813           2.006111          3.254576          2.042884   
min          20.463333          17.200000         28.766667         15.100000   
25%          37.900000          20.790000         36.900000         19.530000   
50%          40.500000          22.100000         38.530000         20.666667   
75%          43.260000          23.290000         41.760000         22.100000   
max          56.026667          29.236000         50.163333         26.200000   

       office_room_hum  bath_room_temp  ...  parent_room_temp  \
count     19735.000000    19735.000000  ...      19735.000000   
mean         39.026904       19.592106  ...         19.485828   
std           4.341321        1.844623  ...          2.014712   
min          27.660000       15.330000  ...         14.890000   
25%          35.530000       18.277500  ...         18.000000   
50%          38.400000       19.390000  ...         19.390000   
75%          42.156667       20.619643  ...         20.600000   
max          51.090000       25.795000  ...         24.500000   

       parent_room_hum  Outside_station_temp   Press_mm_hg  \
count     19735.000000          19735.000000  19735.000000   
mean         41.552401              7.411665    755.522602   
std           4.151497              5.317409      7.399441   
min          29.166667             -5.000000    729.300000   
25%          38.500000              3.666667    750.933333   
50%          40.900000              6.916667    756.100000   
75%          44.338095             10.408333    760.933333   
max          53.326667             26.100000    772.300000   

       outside_station_hum     Windspeed    Visibility     Tdewpoint  \
count         19735.000000  19735.000000  19735.000000  19735.000000   
mean             79.750418      4.039752     38.330834      3.760707   
std              14.901088      2.451221     11.794719      4.194648   
min              24.000000      0.000000      1.000000     -6.600000   
25%              70.333333      2.000000     29.000000      0.900000   
50%              83.666667      3.666667     40.000000      3.433333   
75%              91.666667      5.500000     40.000000      6.566667   
max             100.000000     14.000000     66.000000     15.500000   

                rv1           rv2  
count  19735.000000  19735.000000  
mean      24.988033     24.988033  
std       14.496634     14.496634  
min        0.005322      0.005322  
25%       12.497889     12.497889  
50%       24.897653     24.897653  
75%       37.583769     37.583769  
max       49.996530     49.996530  

[8 rows x 27 columns]

In [9]:
from sklearn.preprocessing import MinMaxScaler # normalize the dataset to be between the range of 0 to 1 
from sklearn.model_selection import train_test_split
Scaler = MinMaxScaler()
normalise_energy_data_revised = pd.DataFrame(Scaler.fit_transform(energy_data_revised), columns=energy_data_revised.columns)
feature_energy_data = normalise_energy_data_revised.drop(columns = ['Appliances']) #drop the target variable

C:\Users\Nichel\Anaconda2\lib\site-packages\sklearn\preprocessing\data.py:334: DataConversionWarning: Data with input dtype int64, float64 were all converted to float64 by MinMaxScaler.
  return self.partial_fit(X, y)


In [10]:
# 
Appliances_target = normalise_energy_data_revised['Appliances']

In [11]:
# Train and test data
x_train, x_test, y_train, y_test = train_test_split(feature_energy_data, Appliances_target, test_size =0.3, random_state =42)
x_train, x_test, y_train, y_test

(       kitchen_temp  kitchen_hum  living_room_temp  living_room_hum  \
 9129       0.497360     0.236767          0.122850         0.565939   
 2453       0.286167     0.482616          0.188999         0.669978   
 9152       0.422386     0.230529          0.057427         0.606430   
 12694      0.560718     0.446840          0.280834         0.704002   
 16952      0.835269     0.422071          1.000000         0.318493   
 360        0.191130     0.421154          0.138115         0.594058   
 5007       0.307286     0.381892          0.213230         0.499578   
 14816      0.633580     0.320338          0.420887         0.441278   
 14538      0.662795     0.217595          0.522656         0.326928   
 289        0.263992     0.377213          0.188999         0.542600   
 19347      0.771911     0.368131          0.399806         0.616834   
 2375       0.221753     0.547014          0.155924         0.739010   
 10877      0.613516     0.339602          0.254422         0.66

In [12]:
from sklearn.linear_model import LinearRegression #build your model
linear_model = LinearRegression()
linear_model.fit(x_train, y_train)

LinearRegression(copy_X=True, fit_intercept=True, n_jobs=None,
         normalize=False)

In [13]:
predicted_value =linear_model.predict(x_test) #predict your model

In [16]:
from sklearn.metrics import mean_absolute_error
mae = round(mean_absolute_error(y_test, predicted_value), 2)
mae

0.05

#### Question 13
What is the Mean Absolute Error?
#### Answer = 0.05

In [17]:
rss = round(np.sum(np.square(y_test - predicted_value)), 2)
rss

45.35

#### Question 14
What is the Residual Sum of Square?
#### Answer = 45.35


In [18]:
mse= round(np.mean((predicted_value - y_test)**2), 3)
mse

0.008

In [22]:
from sklearn.metrics import  mean_squared_error
rmse = round(np.sqrt(mean_squared_error(y_test, predicted_value)), 3)
rmse

0.088

#### Question 15
What is the Root Mean Square Error?

#### Answer = 0.088

In [24]:
from sklearn.metrics import r2_score
r2_score = round(r2_score(y_test, predicted_value), 2)
r2_score

0.15

#### Question  16
What is the Coefficient of Determination?
#### Answer = 0.15

In [25]:
from sklearn.linear_model import Ridge
ridge_reg = Ridge(alpha=0)
ridge_reg.fit(x_train, y_train)

Ridge(alpha=0, copy_X=True, fit_intercept=True, max_iter=None,
   normalize=False, random_state=None, solver='auto', tol=0.001)

In [26]:
from sklearn.linear_model import Lasso
lasso_reg = Lasso(alpha=1)
lasso_reg.fit(x_train, y_train)

Lasso(alpha=1, copy_X=True, fit_intercept=True, max_iter=1000,
   normalize=False, positive=False, precompute=False, random_state=None,
   selection='cyclic', tol=0.0001, warm_start=False)

In [27]:
#comparing the effects of regularisation
def get_weights_df(model, feat, col_name):
  weights = pd.Series(model.coef_, feat.columns).sort_values()
  weights_df = pd.DataFrame(weights).reset_index()
  weights_df.columns = ['Features', col_name]
  weights_df[col_name].round(3)
  return weights_df

linear_model_weights = get_weights_df(linear_model, x_train, 'Linear_Model_Weight')
ridge_weights_df = get_weights_df(ridge_reg, x_train, 'Ridge_Weight')
lasso_weights_df = get_weights_df(lasso_reg, x_train, 'Lasso_weight')

final_weights = pd.merge(linear_model_weights, ridge_weights_df, on='Features')
final_weights = pd.merge(final_weights, lasso_weights_df, on='Features')
final_weights

Features  Linear_Model_Weight  Ridge_Weight  Lasso_weight
0         living_room_hum            -0.456698 -4.509427e-01          -0.0
1    Outside_station_temp            -0.321860 -3.298563e-01           0.0
2        living_room_temp            -0.236178 -2.320746e-01           0.0
3        parent_room_temp            -0.189941 -1.886170e-01           0.0
4      teenager_room2_hum            -0.157595 -1.564342e-01          -0.0
5     outside_station_hum            -0.077671 -8.220193e-02          -0.0
6                    RH_7            -0.044614 -4.609337e-02          -0.0
7         parent_room_hum            -0.039800 -3.829966e-02          -0.0
8          bath_room_temp            -0.015657 -1.730898e-02           0.0
9            kitchen_temp            -0.003281 -3.281051e-03           0.0
10                    rv1             0.000770 -7.332193e+12          -0.0
11                    rv2             0.000770  7.332193e+12          -0.0
12            Press_mm_hg             0.006839  5.107303e-03          -0.0
13      ironing_room_temp             0.010319  1.015236e-02           0.0
14             Visibility             0.012307  1.266377e-02           0.0
15          bath_room_hum             0.016006  1.544017e-02           0.0
16        office_room_hum             0.026386  2.504785e-02           0.0
17       office_room_temp             0.028981  2.914372e-02           0.0
18              Windspeed             0.029183  2.815200e-02           0.0
19   outside_building_hum             0.038049  3.791602e-02          -0.0
20       ironing_room_hum             0.096048  9.656299e-02           0.0
21    teenager_room2_temp             0.101995  1.006521e-01           0.0
22              Tdewpoint             0.117758  1.221082e-01           0.0
23  outside_building_temp             0.236425  2.365719e-01           0.0
24      laundry_room_temp             0.290627  2.893305e-01           0.0
25            kitchen_hum             0.553547  5.510378e-01           0.0

#### Question 17
Obtain the feature weights from your linear model above. which feature has the lowest and highest weights respectively?
#### Answer = RH_2 (kitchen_hum), RH_1 ( living_room_hum)

In [33]:
from sklearn.linear_model import Ridge
ridge_reg = Ridge(alpha=0.04, normalize = True)
ridge_reg.fit(x_train, y_train)

Ridge(alpha=0.04, copy_X=True, fit_intercept=True, max_iter=None,
   normalize=True, random_state=None, solver='auto', tol=0.001)

In [35]:
ridge_reg.score(x_test, y_test)

0.1322685595539379

In [34]:
mse= round(np.mean((predicted_value - y_test)**2), 3)
mse

0.008

In [36]:
rmse = round(np.sqrt(mean_squared_error(y_test, predicted_value)), 3)
rmse

0.088

#### Question 18
Train a ridge regression model with an alpha value of 0.4. Is there any change to the root mean square of error when evaluated on the test set
#### Answer = No

In [30]:
from sklearn.linear_model import Lasso
lasso_reg = Lasso(alpha=0.001)
lasso_reg.fit(x_train, y_train)

Lasso(alpha=0.001, copy_X=True, fit_intercept=True, max_iter=1000,
   normalize=False, positive=False, precompute=False, random_state=None,
   selection='cyclic', tol=0.0001, warm_start=False)

In [31]:
#comparing the effects of regularisation
def get_weights_df(model, feat, col_name):
  weights = pd.Series(model.coef_, feat.columns).sort_values()
  weights_df = pd.DataFrame(weights).reset_index()
  weights_df.columns = ['Features', col_name]
  weights_df[col_name].round(3)
  return weights_df

linear_model_weights = get_weights_df(linear_model, x_train, 'Linear_Model_Weight')
ridge_weights_df = get_weights_df(ridge_reg, x_train, 'Ridge_Weight')
lasso_weights_df = get_weights_df(lasso_reg, x_train, 'Lasso_weight')

final_weights = pd.merge(linear_model_weights, ridge_weights_df, on='Features')
final_weights = pd.merge(final_weights, lasso_weights_df, on='Features')
final_weights

Features  Linear_Model_Weight  Ridge_Weight  Lasso_weight
0         living_room_hum            -0.456698     -0.526358     -0.000000
1    Outside_station_temp            -0.321860     -0.272741      0.000000
2        living_room_temp            -0.236178     -0.316487      0.000000
3        parent_room_temp            -0.189941     -0.218294     -0.000000
4      teenager_room2_hum            -0.157595     -0.169271     -0.000110
5     outside_station_hum            -0.077671     -0.031945     -0.049557
6                    RH_7            -0.044614     -0.057737     -0.000000
7         parent_room_hum            -0.039800     -0.019162     -0.000000
8          bath_room_temp            -0.015657      0.035087     -0.000000
9            kitchen_temp            -0.003281      0.020435      0.000000
10                    rv1             0.000770     -0.005254     -0.000000
11                    rv2             0.000770     -0.005254     -0.000000
12            Press_mm_hg             0.006839     -0.000169     -0.000000
13      ironing_room_temp             0.010319      0.011089     -0.000000
14             Visibility             0.012307      0.006838      0.000000
15          bath_room_hum             0.016006      0.013611      0.000000
16        office_room_hum             0.026386      0.079490      0.000000
17       office_room_temp             0.028981      0.033979     -0.000000
18              Windspeed             0.029183      0.024758      0.002912
19   outside_building_hum             0.038049      0.016295     -0.000000
20       ironing_room_hum             0.096048      0.112152      0.000000
21    teenager_room2_temp             0.101995      0.107629      0.000000
22              Tdewpoint             0.117758      0.071861      0.000000
23  outside_building_temp             0.236425      0.263375      0.000000
24      laundry_room_temp             0.290627      0.273168      0.000000
25            kitchen_hum             0.553547      0.524307      0.017880

#### Question 19
Train a lasso regression model with an alpha value of 0.001 and obtain the new feature weights with it. How many of the features have non zero feature weights
#### Answer = 4